### ISSUE TITLE
Jagged edges in the output the masks

### ISSUE THREAD

**Issue by vanga** [SysCV/transfiner#25]
Hi,

Config: mask_rcnn_R_101_FPN_3x_deform.yaml
checkpoint: output_3x_transfiner_r101_deform.pth

All other settings are defaults from the yaml config.

I am doing prediction like this (I used the demo.py script as wel and the results are same). I don't see any errors/warning in the logs
```
im = cv2.imread(src_img_path)
predictor = DefaultPredictor(cfg)
outputs = predictor(im)
```

I see that the edges of the masks are not smooth even for the sample images in the repo.  Is this expected? I am askng this because when I look at the gifs in the readme, they seem to have smooth edges. so, I was expecting smooth edges.

![000000131444-out](https://user-images.githubusercontent.com/5126396/174301343-ee8bf16a-960d-441c-b9f4-03db75d30175.jpeg)

![000000126137-out](https://user-images.githubusercontent.com/5126396/174301347-d91eb4b6-f8d1-4686-81fe-4b6e447591a3.jpeg)
![000000008844-out](https://user-images.githubusercontent.com/5126396/174301351-bda37ee9-9666-45b0-a03f-586b8d103071.jpeg)

![000000157365-out](https://user-images.githubusercontent.com/5126396/174301339-e9ee97a5-9ad9-4a0c-8b42-a678b9682de6.jpeg)

**Comment by lkeab** [comment]
Hi, have this [line](https://github.com/SysCV/transfiner/blob/main/detectron2/modeling/roi_heads/mask_head.py#L565) triggered during the inference process? You can try to increase the dilation kernel range to reduce the artifacts. This is the result of the woman running on my local machine.
<img width="496" alt="image" src="https://user-images.githubusercontent.com/17427852/174347847-05e7a84c-99db-4d17-9327-6b89b2e8bf0f.png">

**Comment by vanga** [comment]
Hi @lkeab without the background removed, these edges are not noticeable. I don't see the jagged edges even with my inference, atleast they are not noticeable this way I think.
![000000008844](https://user-images.githubusercontent.com/5126396/174417061-20b7c2d6-a0de-4a34-8960-6b1bf6e6b147.jpeg)

I simply do this to remove the background, so I don't think it is an issue with the background removal process.

```
from PIL import Image
img = Image.open(src_img_path)

np_img = np.copy(np.asarray(img))
neg_mask = ~mask
np_img[neg_mask] = 255
plt.imshow(np_img)
```

I wasn't setting VIS_PERIOD if it is not the default. I have set it now and I can confrim that the code is going to dilating. Though, I am not sure if the output is any different.. I also tried with 7x7 kernel.. not much improvement.

![000000008844-out](https://user-images.githubusercontent.com/5126396/174420223-1fd717d7-d988-4975-b638-5a9ca2304a85.jpg)

I see below logs during inference, not sure if they are of any significance

```
fpn input shapes: {'res2': ShapeSpec(channels=256, height=None, width=None, stride=4), 'res3': ShapeSpec(channels=512, height=None, width=None, stride=8), 'res4': ShapeSpec(channels=1024, height=None, width=None, stride=16), 'res5': ShapeSpec(channels=2048, height=None, width=None, stride=32)}
/home/ubuntu/vangap/transfiner/detectron2/structures/image_list.py:99: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  max_size = (max_size + (stride - 1)) // stride * stride
/home/ubuntu/miniconda3/envs/transfiner/lib/python3.7/site-packages/torch/functional.py:568: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  ../aten/src/ATen/native/TensorShape.cpp:2228.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
dilating---------------------------------------
/home/ubuntu/vangap/transfiner/detectron2/modeling/roi_heads/mask_head.py:60: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  dim_t = temperature ** (2 * (dim_t // 2) / num_pos_feats)
  ```

**Comment by lkeab** [comment]
I did't observe this warning when running my code.

**Comment by vanga** [comment]
Could there be an issue with the pytorch/cuda versions?
I am using all the latest versions, and on A100.

**Comment by lkeab** [comment]
Not sure. I am using the NVIDIA TITAN RTX, Pytorch 1.7.1 and  Cuda 11.0.

**Comment by vanga** [comment]
If you don't mind, is it possible to provide mask object for one of these images(and details on which model and setttings are used) or provide an option in huggingspaces to download it?

I would like to compare it with what I am seeing and confirm if it's an env issue or something else.

**Comment by anl13** [comment]
> Hi @lkeab without the background removed, these edges are not noticeable. I don't see the jagged edges even with my inference, atleast they are not noticeable this way I think. ![000000008844](https://user-images.githubusercontent.com/5126396/174417061-20b7c2d6-a0de-4a34-8960-6b1bf6e6b147.jpeg)
> 
> I simply do this to remove the background, so I don't think it is an issue with the background removal process.
> 
> ```
> from PIL import Image
> img = Image.open(src_img_path)
> 
> np_img = np.copy(np.asarray(img))
> neg_mask = ~mask
> np_img[neg_mask] = 255
> plt.imshow(np_img)
> ```
> 
> I wasn't setting VIS_PERIOD if it is not the default. I have set it now and I can confirm that the code is going to `dilating` step. Though, I am not sure if the output is any different.. I also tried with 7x7 kernel.. not much improvement.
> 
> ![000000008844-out](https://user-images.githubusercontent.com/5126396/174420223-1fd717d7-d988-4975-b638-5a9ca2304a85.jpg)
> 
> I see below logs during inference, not sure if they are of any significance
> 
> ```
> fpn input shapes: {'res2': ShapeSpec(channels=256, height=None, width=None, stride=4), 'res3': ShapeSpec(channels=512, height=None, width=None, stride=8), 'res4': ShapeSpec(channels=1024, height=None, width=None, stride=16), 'res5': ShapeSpec(channels=2048, height=None, width=None, stride=32)}
> /home/ubuntu/vangap/transfiner/detectron2/structures/image_list.py:99: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
>   max_size = (max_size + (stride - 1)) // stride * stride
> /home/ubuntu/miniconda3/envs/transfiner/lib/python3.7/site-packages/torch/functional.py:568: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  ../aten/src/ATen/native/TensorShape.cpp:2228.)
>   return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
> dilating---------------------------------------
> /home/ubuntu/vangap/transfiner/detectron2/modeling/roi_heads/mask_head.py:60: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
>   dim_t = temperature ** (2 * (dim_t // 2) / num_pos_feats)
> ```

I encountered the same warnings! My env setting: ubuntu 20.04, python 3.7.13, pytorch 1.11.0, cuda 11.3.

**Comment by vanga** [comment]
@anl13 Do you also see the jagged edges?

**Comment by anl13** [comment]
@vanga Yes, I observed exactly the same results.  In fact, I have removed these warnings by replacing `a//b` with `torch.div(a, b, rounding_mode='floor')`. However, there is no help. See result below. By the way, my GPU for test is Nvidia Titan X. 
![demo](https://user-images.githubusercontent.com/10414372/175886334-56cc3959-be32-4a45-9468-9b6bf5815b7b.png)

**Comment by vanga** [comment]
@lkeab Can we keep this open until thee is a closure? considering it is not just me?

w.r.t checking AP results, could you please share some pointers on how I can do that?

**Comment by lkeab** [comment]
Hi, for checking AP results, you can report the inference results on coco val set using our provided pretrained model, such as R50-FPN-1x or 3X. I will also look into this jagged edge problem to see whether can help fix it. It's not that noticeable when drawing masks directly on images.

**Comment by lkeab** [comment]
Hi, we also updated the code. You can try the latest one to see whether it helps or not.

**Comment by anl13** [comment]
Here are my results with updated code. There are improvements compared with previous code. I test three pretrained models for human segmentation. 

model: `output_3x_transfiner_r101_deform.pth`: 
![output000000008844](https://user-images.githubusercontent.com/10414372/176574498-86ef52c3-fa31-4ca7-a38a-bbeeda543aff.jpg)
![output000000126137](https://user-images.githubusercontent.com/10414372/176574527-d37ebcc5-13c9-49ba-9781-8f807a865e83.jpg)
![output000000131444](https://user-images.githubusercontent.com/10414372/176574536-f68b8b28-417c-4b78-bf01-f1301cd5d47a.jpg)
![output000000157365](https://user-images.githubusercontent.com/10414372/176574548-aed010d0-869b-4b37-b2b2-bbca5b69f159.jpg)

model: `output_3x_transfiner_r101.pth`:
![output2000000008844](https://user-images.githubusercontent.com/10414372/176574646-2142fac9-9701-48d4-b1ca-b168c28085a2.jpg)
![output2000000126137](https://user-images.githubusercontent.com/10414372/176574650-2e9a5513-44e7-4b2b-86e4-3977c205c95a.jpg)
![output2000000131444](https://user-images.githubusercontent.com/10414372/176574653-f20208f9-26ab-4917-bcd2-ae089bd41bdd.jpg)
![output2000000157365](https://user-images.githubusercontent.com/10414372/176574658-85f9da68-00a6-4686-b186-ed129238d78b.jpg)

model: `output_3x_transfiner_r50.pth`: 
![output3000000008844](https://user-images.githubusercontent.com/10414372/176574663-a3378df3-5e4c-44fa-bfe4-bb1a8d29c9f3.jpg)
![output3000000126137](https://user-images.githubusercontent.com/10414372/176574675-c5fd27d9-9a90-464d-9586-5405c4674d88.jpg)
![output3000000131444](https://user-images.githubusercontent.com/10414372/176574685-c9cd89c1-7d76-4f22-ab05-896af15a38cc.jpg)
![output3000000157365](https://user-images.githubusercontent.com/10414372/176574700-2b2a0291-108c-4521-b5d3-25623d472ff6.jpg)
